src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCoalescingAssignmentTests.cs

```diff
@@ -1580,7 +1941,376 @@ static void M(dynamic d1, dynamic d2)
+                // (4,10): error CS0177: The out parameter 'o2' must be assigned to before control leaves the current method
+                //     void M(in object o1, out object o2)
+                Diagnostic(ErrorCode.ERR_ParamUnassigned, "M").WithArguments("o2").WithLocation(4, 10),
+                // (6,9): error CS8331: Cannot assign to variable 'in object' because it is a readonly variable
+                //         o1 ??= null;
+                Diagnostic(ErrorCode.ERR_AssignReadonlyNotField, "o1").WithArguments("variable", "in object").WithLocation(6, 9),
+                // (7,9): error CS0269: Use of unassigned out parameter 'o2'
+                //         o2 ??= null;
+                Diagnostic(ErrorCode.ERR_UseDefViolationOut, "o2").WithArguments("o2").WithLocation(7, 9)
+            });
+        }
+
+        [Fact]
+        public void ObsoleteConversion()
+        {
+            var source = @"
+using System;
+class C
+{
+    void M()
+    {
+        C c = null;
+        c ??= new D();
+        c ??= new E();
+    }
+}
+class D
+{
+    [Obsolete(""D is obsolete"", true)]
+    public static implicit operator C(D d) => null;
+}
+class E
+{
+    [Obsolete(""E is obsolete"", false)]
+    public static implicit operator C(E e) => null;
+}";
+
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics(new DiagnosticDescription[] {
+                // (8,15): error CS0619: 'D.implicit operator C(D)' is obsolete: 'D is obsolete'
+                //         c ??= new D();
+                Diagnostic(ErrorCode.ERR_DeprecatedSymbolStr, "new D()").WithArguments("D.implicit operator C(D)", "D is obsolete").WithLocation(8, 15),
+                // (9,15): warning CS0618: 'E.implicit operator C(E)' is obsolete: 'E is obsolete'
+                //         c ??= new E();
+                Diagnostic(ErrorCode.WRN_DeprecatedSymbolStr, "new E()").WithArguments("E.implicit operator C(E)", "E is obsolete").WithLocation(9, 15)
+            });
+        }
+
+        [Fact]
+        public void DestructuringSyntaxDoesNotWork()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        int? x = null, y = null;
+        (x, y) ??= (1, 2);
+    }
+}";
+
+            CreateCompilation(source).VerifyDiagnostics(new DiagnosticDescription[] {
+                // (7,9): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         (x, y) ??= (1, 2);
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "(x, y)").WithLocation(7, 9)
+            });
+        }
+
+        [Fact]
+        public void UseSiteDiagnostics()
+        {
+            var source = @"
+using ClassAlias = Class1;
+class C : ClassAlias
+{
+    void M()
+    {
+        ClassAlias a = null;
+        a ??= this;
+    }
+}
+";
+
+            // The last diagnostic is the actual use-site diagnostic we are testing. The rest are setup.
```

Comment:
>The last diagnostic is the actual use-site diagnostic we are testing. The rest are setup. [](start = 15, length = 89)

Is there a simple scenario with just a single use-site error?

Perhaps compile c.exe with b.dll but not a.dll:
a.dll:
```
public class A { }
```
b.dll:
```
public class B : A { }
```
c.exe:
```
class C
{
    static void Main()
    {
        var c = new C();
        var b = new B();
        c ??= b;
    }
}
```

Reply:
Yeah. That will have a second error, but it's not a use-site diagnostic, unlike this scenario.

---
In reply to: [216111314](https://github.com/dotnet/roslyn/pull/29691#discussion_r216111314) [](ancestors = 216111314)